File: TarsEngine/Services/PerformanceIssueDetector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using Microsoft.Extensions.Logging;
using TarsEngine.Models;
using TarsEngine.Services.Interfaces;

namespace TarsEngine.Services
{
    /// <summary>
    /// Detects performance issues in code
    /// </summary>
    public class PerformanceIssueDetector(ILogger<PerformanceIssueDetector> logger) : IPerformanceIssueDetector
    {
        private readonly ILogger<PerformanceIssueDetector> _logger = logger;

        /// <inheritdoc/>
        public string Language => "csharp";

        /// <inheritdoc/>
        public CodeIssueType IssueType => CodeIssueType.Performance;

        /// <inheritdoc/>
        public List<CodeIssue> DetectIssues(string content, List<CodeStructure> structures)
        {
            var issues = new List<CodeIssue>();

            try
            {
                if (string.IsNullOrWhiteSpace(content))
                {
                    return issues;
                }

                issues.AddRange(DetectInefficientLoops(content));
                issues.AddRange(DetectLargeObjectCreation(content));
                issues.AddRange(DetectExcessiveMemoryUsage(content));
                issues.AddRange(DetectInefficientStringOperations(content));
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error detecting performance issues in code");
            }

            return issues;
        }

        /// <inheritdoc/>
        public List<CodeIssue> DetectInefficientLoops(string content)
        {
            var issues = new List<CodeIssue>();

            try
            {
                if (string.IsNullOrWhiteSpace(content))
                {
                    return issues;
                }

                // Detect collection modification inside foreach loops
                var foreachModificationRegex = new Regex(@"foreach\s*\([^)]+\)\s*{[^}]*\b(Add|Remove|Clear|Insert)\b", RegexOptions.Compiled);
                var foreachModificationMatches = foreachModificationRegex.Matches(content);
                foreach (Match match in foreachModificationMatches)
                {
                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Performance,
                        Severity = TarsEngine.Models.IssueSeverity.Major,
                        Description = "Use a for loop or ToList() when modifying collections during iteration",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        Title = "Collection Modification in Loop"
                    });
                }

                // Detect inefficient LINQ usage
                var inefficientLinqRegex = new Regex(@"\.Where\([^)]+\)\.Where\(", RegexOptions.Compiled);
                var inefficientLinqMatches = inefficientLinqRegex.Matches(content);
                foreach (Match match in inefficientLinqMatches)
                {
                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Performance,
                        Severity = TarsEngine.Models.IssueSeverity.Minor,
                        Description = "Combine multiple Where() calls into a single call with a compound condition",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        Title = "Multiple Where() Calls"
                    });
                }

                // Detect inefficient loop variable usage
                var loopVariableRegex = new Regex(@"for\s*\([^;]+;\s*[^;]+;\s*[^)]+\)\s*{[^}]*\b(Length|Count)\b", RegexOptions.Compiled);
                var loopVariableMatches = loopVariableRegex.Matches(content);
                foreach (Match match in loopVariableMatches)
                {
                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Performance,
                        Severity = TarsEngine.Models.IssueSeverity.Trivial,
                        Description = "Cache the collection length/count in a variable before the loop",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        Title = "Uncached Collection Length"
                    });
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error detecting inefficient loops");
            }

            return issues;
        }

        /// <inheritdoc/>
        public List<CodeIssue> DetectLargeObjectCreation(string content)
        {
            var issues = new List<CodeIssue>();

            try
            {
                if (string.IsNullOrWhiteSpace(content))
                {
                    return issues;
                }

                // Detect large array creation
                var largeArrayRegex = new Regex(@"new\s+[a-zA-Z0-9_<>]+\s*\[\s*(\d+)\s*\]", RegexOptions.Compiled);
                var largeArrayMatches = largeArrayRegex.Matches(content);
                foreach (Match match in largeArrayMatches)
                {
                    if (match.Groups.Count > 1 && int.TryParse(match.Groups[1].Value, out int size) && size > 1000000)
                    {
                        var lineNumber = GetLineNumber(content, match.Index);
                        issues.Add(new CodeIssue
                        {
                            Type = CodeIssueType.Performance,
                            Severity = TarsEngine.Models.IssueSeverity.Major,
                            Description = $"Consider using a more memory-efficient data structure or streaming approach for large array (size: {size})",
                            Location = new CodeLocation
                            {
                                StartLine = lineNumber,
                                EndLine = lineNumber
                            },
                            Title = "Large Array Creation"
                        });
                    }
                }

                // Detect large collection initializers
                var largeCollectionRegex = new Regex(@"new\s+[a-zA-Z0-9_<>]+\s*\{([^}]{1000,})\}", RegexOptions.Compiled);
                var largeCollectionMatches = largeCollectionRegex.Matches(content);
                foreach (Match match in largeCollectionMatches)
                {
                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Performance,
                        Severity = TarsEngine.Models.IssueSeverity.Minor,
                        Description = "Consider initializing the collection incrementally or using a factory method",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        Title = "Large Collection Initializer"
                    });
                }

                // Detect large string concatenation
                var largeStringConcatRegex = new Regex(@"string\s+[a-zA-Z0-9_]+\s*=\s*([^;]+\+[^;]+){5,}", RegexOptions.Compiled);
                var largeStringConcatMatches = largeStringConcatRegex.Matches(content);
                foreach (Match match in largeStringConcatMatches)
                {
                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Performance,
                        Severity = TarsEngine.Models.IssueSeverity.Minor,
                        Description = "Use StringBuilder for multiple string concatenations",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        Title = "Multiple String Concatenations"
                    });
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error detecting large object creation");
            }

            return issues;
        }

        /// <inheritdoc/>
        public List<CodeIssue> DetectExcessiveMemoryUsage(string content)
        {
            var issues = new List<CodeIssue>();

            try
            {
                if (string.IsNullOrWhiteSpace(content))
                {
                    return issues;
                }

                // Detect missing using statements for disposable objects
                var disposableRegex = new Regex(@"new\s+(FileStream|StreamReader|StreamWriter|SqlConnection|SqlCommand|MemoryStream|NetworkStream)\s*\([^)]*\)(?!\s*using|\s*=\s*null|\s*;\s*using|\s*;\s*try)", RegexOptions.Compiled);
                var disposableMatches = disposableRegex.Matches(content);
                foreach (Match match in disposableMatches)
                {
                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Performance,
                        Severity = TarsEngine.Models.IssueSeverity.Critical,
                        Description = "Disposable object created without using statement or proper disposal",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        CodeSnippet = match.Value,
                        SuggestedFix = "Use a using statement or using declaration to ensure proper disposal"
                    });
                }

                // Detect large static collections
                var staticCollectionRegex = new Regex(@"static\s+[a-zA-Z0-9_<>]+\s+[a-zA-Z0-9_]+\s*=\s*new\s+[a-zA-Z0-9_<>]+", RegexOptions.Compiled);
                var staticCollectionMatches = staticCollectionRegex.Matches(content);
                foreach (Match match in staticCollectionMatches)
                {
                    if (match.Value.Contains("List<") || match.Value.Contains("Dictionary<") ||
                        match.Value.Contains("HashSet<") || match.Value.Contains("Collection<"))
                    {
                        var lineNumber = GetLineNumber(content, match.Index);
                        issues.Add(new CodeIssue
                        {
                            Type = CodeIssueType.Performance,
                            Severity = TarsEngine.Models.IssueSeverity.Major,
                            Description = "Static collection detected",
                            Location = new CodeLocation
                            {
                                StartLine = lineNumber,
                                EndLine = lineNumber
                            },
                            CodeSnippet = match.Value,
                            SuggestedFix = "Consider using a non-static collection or implementing a cache with size limits"
                        });
                    }
                }

                // Detect excessive boxing/unboxing
                var boxingRegex = new Regex(@"(object)\s+[a-zA-Z0-9_]+\s*=\s*\d+|object\s*\[\s*\]|ArrayList|Hashtable", RegexOptions.Compiled);
                var boxingMatches = boxingRegex.Matches(content);
                foreach (Match match in boxingMatches)
                {
                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Performance,
                        Severity = TarsEngine.Models.IssueSeverity.Major,
                        Description = "Potential boxing/unboxing operation detected",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        CodeSnippet = match.Value,
                        SuggestedFix = "Use generic collections and avoid unnecessary conversions between value types and object"
                    });
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error detecting excessive memory usage");
            }

            return issues;
        }

        /// <inheritdoc/>
        public List<CodeIssue> DetectInefficientStringOperations(string content)
        {
            var issues = new List<CodeIssue>();

            try
            {
                if (string.IsNullOrWhiteSpace(content))
                {
                    return issues;
                }

                // Detect string concatenation in loops
                var stringConcatLoopRegex = new Regex(@"(for|foreach|while)[^{]*{[^}]*\+\=\s*[""']", RegexOptions.Compiled);
                var stringConcatLoopMatches = stringConcatLoopRegex.Matches(content);
                foreach (Match match in stringConcatLoopMatches)
                {
                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Performance,
                        Severity = TarsEngine.Models.IssueSeverity.Critical,
                        Description = "String concatenation in loop detected",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        CodeSnippet = match.Value,
                        SuggestedFix = "Use StringBuilder for string concatenation in loops"
                    });
                }

                // Detect inefficient string.Substring usage
                var substringRegex = new Regex(@"\.Substring\s*\(\s*0\s*,", RegexOptions.Compiled);
                var substringMatches = substringRegex.Matches(content);
                foreach (Match match in substringMatches)
                {
                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Performance,
                        Severity = TarsEngine.Models.IssueSeverity.Minor,
                        Description = "Inefficient Substring usage detected",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        CodeSnippet = match.Value,
                        SuggestedFix = "Use string[..end] range operator instead of Substring(0, end)"
                    });
                }

                // Detect repeated string.Replace calls
                var replaceRegex = new Regex(@"\.Replace\s*\([^)]+\)\.Replace\s*\(", RegexOptions.Compiled);
                var replaceMatches = replaceRegex.Matches(content);
                foreach (Match match in replaceMatches)
                {
                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Performance,
                        Severity = TarsEngine.Models.IssueSeverity.Major,
                        Description = "Multiple string.Replace calls in chain detected",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        CodeSnippet = match.Value,
                        SuggestedFix = "Consider using Regex.Replace for multiple replacements or a StringBuilder"
                    });
                }

                // Detect inefficient string comparison
                var comparisonRegex = new Regex(@"\.ToLower\s*\(\s*\)\s*==|\.ToUpper\s*\(\s*\)\s*==", RegexOptions.Compiled);
                var comparisonMatches = comparisonRegex.Matches(content);
                foreach (Match match in comparisonMatches)
                {
                    var lineNumber = GetLineNumber(content, match.Index);
                    issues.Add(new CodeIssue
                    {
                        Type = CodeIssueType.Performance,
                        Severity = TarsEngine.Models.IssueSeverity.Minor,
                        Description = "Case-insensitive string comparison using ToLower()/ToUpper() detected",
                        Location = new CodeLocation
                        {
                            StartLine = lineNumber,
                            EndLine = lineNumber
                        },
                        CodeSnippet = match.Value,
                        SuggestedFix = "Use string.Equals(s1, s2, StringComparison.OrdinalIgnoreCase) for case-insensitive comparison"
                    });
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error detecting inefficient string operations");
            }

            return issues;
        }

        /// <inheritdoc/>
        public Dictionary<TarsEngine.Services.Interfaces.IssueSeverity, string> GetAvailableSeverities()
        {
            return new Dictionary<TarsEngine.Services.Interfaces.IssueSeverity, string>
            {
                { TarsEngine.Services.Interfaces.IssueSeverity.Critical, "Critical performance issue that causes significant slowdowns" },
                { TarsEngine.Services.Interfaces.IssueSeverity.Major, "High-impact performance issue that should be fixed soon" },
                { TarsEngine.Services.Interfaces.IssueSeverity.Minor, "Medium-impact performance issue" },
                { TarsEngine.Services.Interfaces.IssueSeverity.Trivial, "Low-impact performance issue" },
                { TarsEngine.Services.Interfaces.IssueSeverity.Warning, "Informational performance suggestion" }
            };
        }

        /// <inheritdoc/>
        public int GetLineNumber(string content, int position)
        {
            if (string.IsNullOrEmpty(content) || position < 0 || position >= content.Length)
            {
                return 0;
            }

            // Count newlines before the position
            return content[..position].Count(c => c == '\n') + 1;
        }
    }
}

File: TarsEngine.FSharp.Core/Analysis/PerformanceIssueDetector.fs
namespace TarsEngine.FSharp.Core.Analysis

open System
open System.Text.RegularExpressions
open Microsoft.Extensions.Logging

/// <summary>
/// Implementation of IPerformanceIssueDetector for C# and F# languages.
/// </summary>
type PerformanceIssueDetector(logger: ILogger<PerformanceIssueDetector>) =
    
    /// <summary>
    /// Gets the language supported by this detector.
    /// </summary>
    member _.Language = "csharp"
    
    /// <summary>
    /// Detects issues in the provided content.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectIssues(content: string) =
        try
            logger.LogInformation("Detecting performance issues")
            
            // Detect inefficient loops
            let inefficientLoopIssues = this.DetectInefficientLoops(content)
            
            // Detect large object creation
            let largeObjectCreationIssues = this.DetectLargeObjectCreation(content)
            
            // Detect excessive memory usage
            let excessiveMemoryUsageIssues = this.DetectExcessiveMemoryUsage(content)
            
            // Detect inefficient string operations
            let inefficientStringOperationsIssues = this.DetectInefficientStringOperations(content)
            
            // Combine all issues
            List.concat [
                inefficientLoopIssues
                largeObjectCreationIssues
                excessiveMemoryUsageIssues
                inefficientStringOperationsIssues
            ]
        with
        | ex ->
            logger.LogError(ex, "Error detecting performance issues")
            []
    
    /// <summary>
    /// Detects issues in a file.
    /// </summary>
    /// <param name="filePath">The path to the file.</param>
    /// <returns>A list of detected issues.</returns>
    member this.DetectIssuesInFile(filePath: string) =
        try
            logger.LogInformation("Detecting performance issues in file: {FilePath}", filePath)
            
            // Read the file content
            let content = System.IO.File.ReadAllText(filePath)
            
            // Detect issues in the content
            let issues = this.DetectIssues(content)
            
            // Update issues with file path
            issues |> List.map (fun issue -> { issue with FilePath = Some filePath })
        with
        | ex ->
            logger.LogError(ex, "Error detecting performance issues in file: {FilePath}", filePath)
            []
    
    /// <summary>
    /// Gets the supported issue types.
    /// </summary>
    /// <returns>A list of supported issue types.</returns>
    member _.GetSupportedIssueTypes() =
        [
            CodeIssueType.Performance
        ]
    
    /// <summary>
    /// Detects inefficient loops.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectInefficientLoops(content: string) =
        try
            // Define patterns for inefficient loops
            let patterns = [
                // C# patterns
                @"for\s*\(\s*int\s+i\s*=\s*0\s*;\s*i\s*<\s*([a-zA-Z0-9_\.]+)\.Count\s*;\s*i\s*\+\+\s*\)", "Inefficient loop using Count property in each iteration"
                @"for\s*\(\s*int\s+i\s*=\s*0\s*;\s*i\s*<\s*([a-zA-Z0-9_\.]+)\.Length\s*;\s*i\s*\+\+\s*\)", "Inefficient loop using Length property in each iteration"
                @"foreach\s*\(\s*var\s+[a-zA-Z0-9_]+\s+in\s+([a-zA-Z0-9_\.]+)\.Where\s*\(", "Inefficient use of LINQ Where in foreach loop"
                @"foreach\s*\(\s*var\s+[a-zA-Z0-9_]+\s+in\s+([a-zA-Z0-9_\.]+)\.Select\s*\(", "Inefficient use of LINQ Select in foreach loop"
                @"foreach\s*\(\s*var\s+[a-zA-Z0-9_]+\s+in\s+([a-zA-Z0-9_\.]+)\.OrderBy\s*\(", "Inefficient use of LINQ OrderBy in foreach loop"
                @"foreach\s*\(\s*var\s+[a-zA-Z0-9_]+\s+in\s+([a-zA-Z0-9_\.]+)\.OrderByDescending\s*\(", "Inefficient use of LINQ OrderByDescending in foreach loop"
                
                // F# patterns
                @"for\s+i\s*=\s*0\s+to\s+([a-zA-Z0-9_\.]+)\.Length\s*-\s*1\s+do", "Inefficient loop using Length property in each iteration"
                @"for\s+i\s*=\s*0\s+to\s+([a-zA-Z0-9_\.]+)\.Count\s*-\s*1\s+do", "Inefficient loop using Count property in each iteration"
                @"for\s+[a-zA-Z0-9_]+\s+in\s+([a-zA-Z0-9_\.]+)\s+\|\>\s+Seq\.filter", "Inefficient use of Seq.filter in for loop"
                @"for\s+[a-zA-Z0-9_]+\s+in\s+([a-zA-Z0-9_\.]+)\s+\|\>\s+Seq\.map", "Inefficient use of Seq.map in for loop"
                @"for\s+[a-zA-Z0-9_]+\s+in\s+([a-zA-Z0-9_\.]+)\s+\|\>\s+Seq\.sortBy", "Inefficient use of Seq.sortBy in for loop"
                @"for\s+[a-zA-Z0-9_]+\s+in\s+([a-zA-Z0-9_\.]+)\s+\|\>\s+Seq\.sortByDescending", "Inefficient use of Seq.sortByDescending in for loop"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.IgnoreCase ||| RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    let collectionName = if m.Groups.Count > 1 then m.Groups.[1].Value else ""
                    
                    let suggestedFix = 
                        if message.Contains("Count") || message.Contains("Length") then
                            "Cache the collection length/count before the loop"
                        elif message.Contains("LINQ") || message.Contains("Seq.") then
                            "Materialize the collection before the loop using ToList() or ToArray()"
                        else
                            "Optimize the loop for better performance"
                    
                    {
                        IssueType = CodeIssueType.Performance
                        Severity = IssueSeverity.Warning
                        Message = $"Potential performance issue: {message} for collection '{collectionName}'"
                        LineNumber = Some lineNumber
                        ColumnNumber = Some columnNumber
                        FilePath = None
                        CodeSnippet = Some codeSnippet
                        SuggestedFix = Some suggestedFix
                        AdditionalInfo = Map.empty
                    }
                )
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting inefficient loops")
            []
    
    /// <summary>
    /// Detects large object creation.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectLargeObjectCreation(content: string) =
        try
            // Define patterns for large object creation
            let patterns = [
                // C# patterns
                @"new\s+([a-zA-Z0-9_]+)\s*\[\s*(\d+)\s*\]", "Large array creation"
                @"new\s+List<[a-zA-Z0-9_]+>\s*\(\s*(\d+)\s*\)", "Large list creation"
                @"new\s+Dictionary<[a-zA-Z0-9_,\s]+>\s*\(\s*(\d+)\s*\)", "Large dictionary creation"
                @"new\s+HashSet<[a-zA-Z0-9_]+>\s*\(\s*(\d+)\s*\)", "Large hash set creation"
                @"new\s+StringBuilder\s*\(\s*(\d+)\s*\)", "Large string builder creation"
                
                // F# patterns
                @"Array\.create\s+(\d+)", "Large array creation"
                @"Array\.zeroCreate\s+(\d+)", "Large array creation"
                @"Array\.init\s+(\d+)", "Large array creation"
                @"List\.init\s+(\d+)", "Large list creation"
                @"Array\.replicate\s+(\d+)", "Large array creation"
                @"List\.replicate\s+(\d+)", "Large list creation"
                @"Dictionary\(\s*(\d+)\s*\)", "Large dictionary creation"
                @"HashSet\(\s*(\d+)\s*\)", "Large hash set creation"
                @"StringBuilder\(\s*(\d+)\s*\)", "Large string builder creation"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.IgnoreCase ||| RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    let sizeStr = if m.Groups.Count > 1 then m.Groups.[1].Value else ""
                    
                    // Parse the size
                    let size = 
                        match Int32.TryParse(sizeStr) with
                        | true, value -> value
                        | _ -> 0
                    
                    // Only report issues for large objects
                    if size > 1000000 then
                        Some {
                            IssueType = CodeIssueType.Performance
                            Severity = IssueSeverity.Warning
                            Message = $"Potential performance issue: {message} with size {size:N0}"
                            LineNumber = Some lineNumber
                            ColumnNumber = Some columnNumber
                            FilePath = None
                            CodeSnippet = Some codeSnippet
                            SuggestedFix = Some "Consider using a smaller initial size or a different data structure"
                            AdditionalInfo = Map.empty
                        }
                    else
                        None
                )
                |> Seq.choose id
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting large object creation")
            []
    
    /// <summary>
    /// Detects excessive memory usage.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectExcessiveMemoryUsage(content: string) =
        try
            // Define patterns for excessive memory usage
            let patterns = [
                // C# patterns
                @"\.ToList\(\)\.ToArray\(\)", "Unnecessary conversion between collections"
                @"\.ToArray\(\)\.ToList\(\)", "Unnecessary conversion between collections"
                @"\.AsEnumerable\(\)\.ToList\(\)", "Unnecessary conversion between collections"
                @"\.AsEnumerable\(\)\.ToArray\(\)", "Unnecessary conversion between collections"
                @"\.Select\([^)]+\)\.Select\([^)]+\)", "Chained LINQ operations without materialization"
                @"\.Where\([^)]+\)\.Where\([^)]+\)", "Chained LINQ operations without materialization"
                @"\.OrderBy\([^)]+\)\.OrderBy\([^)]+\)", "Chained LINQ operations without materialization"
                
                // F# patterns
                @"\|\>\s+Seq\.toList\s+\|\>\s+Seq\.toArray", "Unnecessary conversion between collections"
                @"\|\>\s+Seq\.toArray\s+\|\>\s+Seq\.toList", "Unnecessary conversion between collections"
                @"\|\>\s+Seq\.map\s+[^|]+\|\>\s+Seq\.map", "Chained sequence operations without materialization"
                @"\|\>\s+Seq\.filter\s+[^|]+\|\>\s+Seq\.filter", "Chained sequence operations without materialization"
                @"\|\>\s+Seq\.sortBy\s+[^|]+\|\>\s+Seq\.sortBy", "Chained sequence operations without materialization"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.IgnoreCase ||| RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    
                    let suggestedFix = 
                        if message.Contains("conversion") then
                            "Use a single collection type instead of converting between types"
                        elif message.Contains("Chained") then
                            "Materialize intermediate results using ToList() or ToArray()"
                        else
                            "Optimize memory usage"
                    
                    {
                        IssueType = CodeIssueType.Performance
                        Severity = IssueSeverity.Warning
                        Message = $"Potential performance issue: {message}"
                        LineNumber = Some lineNumber
                        ColumnNumber = Some columnNumber
                        FilePath = None
                        CodeSnippet = Some codeSnippet
                        SuggestedFix = Some suggestedFix
                        AdditionalInfo = Map.empty
                    }
                )
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting excessive memory usage")
            []
    
    /// <summary>
    /// Detects inefficient string operations.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of detected issues.</returns>
    member _.DetectInefficientStringOperations(content: string) =
        try
            // Define patterns for inefficient string operations
            let patterns = [
                // C# patterns
                @"string\.Concat\s*\(\s*[^)]+\s*\+\s*[^)]+\s*\)", "Inefficient string concatenation in String.Concat"
                @"for\s*\([^)]+\)\s*\{[^}]*\+=[^}]*\}", "String concatenation in a loop"
                @"while\s*\([^)]+\)\s*\{[^}]*\+=[^}]*\}", "String concatenation in a loop"
                @"foreach\s*\([^)]+\)\s*\{[^}]*\+=[^}]*\}", "String concatenation in a loop"
                @"do\s*\{[^}]*\+=[^}]*\}\s*while", "String concatenation in a loop"
                
                // F# patterns
                @"String\.Concat\s*\(\s*[^)]+\s*\+\s*[^)]+\s*\)", "Inefficient string concatenation in String.Concat"
                @"for\s+[^d]+do[^d]*\+[^d]*done", "String concatenation in a loop"
                @"while\s+[^d]+do[^d]*\+[^d]*done", "String concatenation in a loop"
            ]
            
            // Detect issues using patterns
            patterns
            |> List.collect (fun (pattern, message) ->
                Regex.Matches(content, pattern, RegexOptions.IgnoreCase ||| RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m ->
                    let lineNumber = content.Substring(0, m.Index).Split('\n').Length
                    let columnNumber = m.Index - content.Substring(0, m.Index).LastIndexOf('\n') - 1
                    let codeSnippet = m.Value
                    
                    let suggestedFix = 
                        if message.Contains("loop") then
                            "Use StringBuilder instead of string concatenation in loops"
                        else
                            "Use StringBuilder or string.Format instead of string concatenation"
                    
                    {
                        IssueType = CodeIssueType.Performance
                        Severity = IssueSeverity.Warning
                        Message = $"Potential performance issue: {message}"
                        LineNumber = Some lineNumber
                        ColumnNumber = Some columnNumber
                        FilePath = None
                        CodeSnippet = Some codeSnippet
                        SuggestedFix = Some suggestedFix
                        AdditionalInfo = Map.empty
                    }
                )
                |> Seq.toList
            )
        with
        | ex ->
            logger.LogError(ex, "Error detecting inefficient string operations")
            []
    
    interface IIssueDetector with
        member this.Language = this.Language
        member this.DetectIssues(content) = this.DetectIssues(content)
        member this.DetectIssuesInFile(filePath) = this.DetectIssuesInFile(filePath)
        member this.GetSupportedIssueTypes() = this.GetSupportedIssueTypes()
    
    interface IPerformanceIssueDetector with
        member this.DetectInefficientLoops(content) = this.DetectInefficientLoops(content)
        member this.DetectLargeObjectCreation(content) = this.DetectLargeObjectCreation(content)
        member this.DetectExcessiveMemoryUsage(content) = this.DetectExcessiveMemoryUsage(content)
        member this.DetectInefficientStringOperations(content) = this.DetectInefficientStringOperations(content)
